File: 2022/13.hs
{-# OPTIONS_GHC -Wno-tabs #-}
import Data.List
import Control.Exception
import Utils

data Packet = ListVal [Packet] | IntVal Integer deriving Eq

getInput :: IO [Packet]
getInput = map read <$> filter (not.null) <$> lines <$> readFile "13.txt"

instance (Read Packet) where
	readsPrec d r = [(ListVal x, rest) | (x, rest) <- readsPrec d r] ++ [(IntVal x, rest) | (x, rest) <- readsPrec d r]

instance (Show Packet) where
	showsPrec d (ListVal x) = showsPrec d x
	showsPrec d (IntVal x) = showsPrec d x

instance (Ord Packet) where
	compare (IntVal a) (IntVal b) = compare a b
	compare (ListVal a) (ListVal b) = compare a b
	compare a@(IntVal _) (ListVal b) = compare [a] b
	compare (ListVal a) b@(IntVal _) = compare a [b]

partA ps = sum [i | (i, [a, b]) <- zip [1..] $ chunk 2 ps, a < b]
partB ps = (index1 + 1) * (index2 + 1)
	where
		sentinel1 = ListVal [ListVal [IntVal 2]]
		sentinel2 = ListVal [ListVal [IntVal 6]]
		sorted = sort $ [sentinel1, sentinel2] ++ ps
		Just index1 = elemIndex sentinel1 sorted
		Just index2 = elemIndex sentinel2 sorted

tests :: IO ()
tests = do
	check $ partA testData == 13
	check $ partB testData == 140
	where
		testData = map read ["[1,1,3,1,1]", "[1,1,5,1,1]", "[[1],[2,3,4]]", "[[1],4]", "[9]", "[[8,7,6]]", "[[4,4],4,4]", "[[4,4],4,4,4]", "[7,7,7,7]", "[7,7,7]", "[]", "[3]", "[[[]]]", "[[]]", "[1,[2,[3,[4,[5,6,7]]]],8,9]", "[1,[2,[3,[4,[5,6,0]]]],8,9]"]
		check True = return ()
		check False = throwIO $ AssertionFailed "test failed"

main :: IO ()
main = do
	tests
	dat <- getInput
	print $ partA dat
	print $ partB dat

File: 2022/13.py
#!/usr/bin/python3
from functools import cmp_to_key
dat = [
	([[1,[0,3,5,[2,1,3,3,5]],4,[[],5]],[],[0,[7,[5],7,7]]],[[[],[[],[5,2,8,9,7],1,5],[3,[]]]]),([[[[1,3,6],[7,9,2,7],[5,0,5,8,4]]],[[8,2,6,[]],[],1,6]],[[5,10,4],[6,9,8],[5,[4,[3,8,1,5,1],[4],6,[3]],[[0]],[10],[[5,7],[5,1,1,4,7],[]]],[[[9,7,5,7],[5,1,7,3,1],[8,4,2]],5,[[8,9,3,4,4],4,[2,4,2],3,[6,10,5,7]],9],[[1]]]),[...snip...]
]
def compare(a,b):
	if isinstance(a, int) and isinstance(b, int):
		return a - b
	elif isinstance(a, list) and isinstance(b, list):
		for x, y in zip(a, b):
			ret = compare(x, y)
			if ret != 0:
				return ret
		return len(a) - len(b)
	elif isinstance(a, list):
		return compare(a, [b])
	elif isinstance(b, list):
		return compare([a], b)

print(sum(i for i, (a, b) in enumerate(dat, 1) if compare(a, b) < 0))

packets = [ [[2]], [[6]] ]
for a, b in dat:
	packets.append(a)
	packets.append(b)
packets.sort(key=cmp_to_key(compare))
a = packets.index([[2]]) + 1
b = packets.index([[6]]) + 1
print(a * b)
